_posts/2023-06-13-http4s-authentication-part2.md

```diff
@@ -0,0 +1,894 @@
+end MyDatabase
+```
+
+The `database` is a mutable Map of type `mutable.Map[Int, User]` containing a single user `johnDoe` assigned to an `id` of 1.
+
+Next, we will add two functions to fetch the user's email and counter value given an `id`.
+
+```scala
+trait MyDatabase:
+    ...
+    def getCounterValue (id: Int, database: mutable.Map[Int, User] = database): Long =
+            database(id).counter
+
+    def getEmail(id: Int, database: mutable.Map[Int, User] = database) =
+        database(id).email
+end MyDatabase
+```
+
+The `getCounterValue()` and `getEmail()` functions take an `id` and an optional argument assigned to the `database`, they both use the `id` as a key to fetch the user's `counter` and `email` value respectfully.
+
+```scala
+trait MyDatabase:
+    ...
+    def incrementCounter (id: Int, database: mutable.Map[Int, User] = database) =
+        val counter: Long = database(id).counter + 1
+        database.update(id, User(database(id).username, database(id).email, database(id).password, counter))
+end MyDatabase
+```
+
+The `incrementCounter()` will be used in the case of `Hotp` to increment the counter value by one after we successfully verify a One Time Password.
+
+Here's the full code for MyDatabase.scala
+
+```scala
+package com.rockthejvm
+
+import scala.collection.mutable
+
+case class User(username: String, email: String, password: String, counter: Long)
+
+trait MyDatabase:
+    val database: mutable.Map[Int, User] = mutable.Map(1 -> User("johnDoe", "john@email.com", "password", 1))
+
+    def getCounterValue (id: Int, database: mutable.Map[Int, User] = database): Long =
+        database(id).counter
+
+    def getEmail(id: Int, database: mutable.Map[Int, User] = database) =
+        database(id).email
+
+    def incrementCounter (id: Int, database: mutable.Map[Int, User] = database) =
+        val counter: Long = database(id).counter + 1
+        database.update(id, User(database(id).username, database(id).email, database(id).password, counter))
+end MyDatabase
+```
+
+### 3.1.1 Creating the BarCodeService
+
+Google Authenticator gives us two options to capture the OTP code, we can type it manually or scan a bar code. In this section, we will use the Google Zxing library to create barcode images for our application. Let's create a new scala file called `BarCodeService.scala` where we will add our trait.
+
+```scala
+package com.rockthejvm
+
+import scala.util.*
+import java.net.URLEncoder
+
+trait BarCodeService extends MyDatabase:
+    def getGoogleAuthenticatorBarCode(secretKey: String, account: String, issuer: String, generator: String): Try[String] =
+        generator match
+            case "Hotp" =>
+                Try{
+                s"otpauth://${generator.toLowerCase()}/" +
+                URLEncoder.encode(issuer + ":" + account, "UTF-8").replace("+","%20") +
+                "?secret=" + URLEncoder.encode(secretKey, "UTF-8").replace("+","%20") +
```

Comment:
Is the secret supposed to be exposed like that? I imagine someone can self-sign these OTPs with all the details exposed

Reply:
Put the secret in a private val within OtpService, and only call this function from that class.